src/prof_recent.c

```diff
@@ -0,0 +1,470 @@
+		malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+		reserve = (prof_recent_t *)iallocztm(tsd_tsdn(tsd),
+		    sizeof(prof_recent_t), sz_size2index(sizeof(prof_recent_t)),
+		    false, NULL, true, arena_get(tsd_tsdn(tsd), 0, false),
+		    true);
+		malloc_mutex_lock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	}
+
+	if (prof_recent_alloc_max_load(tsd) == 0) {
+		assert(prof_recent_alloc_count == 0);
+		goto label_rollback;
+	}
+
+	assert(prof_recent_alloc_dummy != NULL);
+	{
+		/* Fill content into the dummy node. */
+		prof_recent_t *node = prof_recent_alloc_dummy;
+		node->usize = usize;
+		nstime_copy(&node->alloc_time,
+		    extent_prof_alloc_time_get(extent));
+		node->alloc_tctx = tctx;
+		extent_prof_recent_alloc_set(tsd, extent, node);
+		nstime_init_zero(&node->dalloc_time);
+		node->dalloc_tctx = NULL;
+	}
+
+	prof_tctx_t *old_alloc_tctx, *old_dalloc_tctx;
+	if (prof_recent_alloc_count == prof_recent_alloc_max_load(tsd)) {
+		/* If upper limit is reached, simply shift the dummy. */
+		assert(prof_recent_alloc_max_load(tsd) != -1);
+		assert_prof_recent_alloc_non_empty(tsd);
+		prof_recent_alloc_dummy = prof_recent_alloc_dummy->next;
+		old_alloc_tctx = prof_recent_alloc_dummy->alloc_tctx;
+		assert(old_alloc_tctx != NULL);
+		old_dalloc_tctx = prof_recent_alloc_dummy->dalloc_tctx;
+		prof_recent_alloc_evict_extent(tsd, prof_recent_alloc_dummy);
+	} else {
+		/* Otherwise use the new node as the dummy. */
+		assert(prof_recent_alloc_max_load(tsd) == -1 ||
+		    prof_recent_alloc_count < prof_recent_alloc_max_load(tsd));
+		if (reserve == NULL) {
+			goto label_rollback;
+		}
+		reserve->next = prof_recent_alloc_dummy->next;
+		prof_recent_alloc_dummy->next = reserve;
+		prof_recent_alloc_dummy = reserve;
+		reserve = NULL;
+		old_alloc_tctx = NULL;
+		old_dalloc_tctx = NULL;
+		++prof_recent_alloc_count;
+	}
+
+	assert(prof_recent_alloc_max_load(tsd) == -1 ||
+	    (prof_recent_alloc_max_load(tsd) > 0 &&
+	    prof_recent_alloc_count <= prof_recent_alloc_max_load(tsd)));
+
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+
+	if (reserve != NULL) {
+		idalloctm(tsd_tsdn(tsd), reserve, NULL, NULL, true, true);
+	}
+
+	/*
+	 * Asynchronously handle the tctx of the old node, so that there's no
+	 * simultaneous holdings of prof_recent_alloc_mtx and tdata->lock.
+	 * In the worst case this may delay the tctx release but it's better
+	 * than holding prof_recent_alloc_mtx for longer.
+	 */
+	if (old_alloc_tctx != NULL) {
+		decrement_recent_count(tsd, old_alloc_tctx);
+	}
+	if (old_dalloc_tctx != NULL) {
+		decrement_recent_count(tsd, old_dalloc_tctx);
+	}
+	return;
+
+label_rollback:
+	assert(extent_prof_recent_alloc_get(tsd, extent) == NULL);
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_recent_alloc_mtx);
+	if (reserve != NULL) {
+		idalloctm(tsd_tsdn(tsd), reserve, NULL, NULL, true, true);
+	}
+	decrement_recent_count(tsd, tctx);
+}
+
+ssize_t
+prof_recent_alloc_max_get() {
```

Comment:
nit: rename to prof_recent_alloc_max_ctl_read() and _write().

Reply:
Also changed the raw getter from `_load()` to `_get()` (because the only reason why I called it `_load()` was to avoid the conflict with this now `prof_recent_alloc_max_ctl_read()` function).